### PR TITLE
Updated Satalia SolveEngine integration

### PR DESCRIPTION
**Pull request by JackDunnNZ**
Ref #11 

@Tapanito, I finally got the chance to look over the integration code! Thanks very much for your patience. I opened a new PR here just to avoid the hassle of pushing to your branch.

Most of the changes are cosmetic only, just tidying spacing etc. I renamed some things to better fit with our existing names too (e.g. seApiDialog -> dialogSolveEngineApi).

Actual changes:

- Using UserCache rather than DocumentCache to save the API key, that way the key is attached to the user in any document they open
- Adding support for resuming an incomplete solve so that an existing model can be reused if a solve fails, e.g. if the API key is entered wrong

There's one more change I'd like to make before we land this, and that's to get rid of `solve_result_num` as an intermediate solve status (since this just how AMPL/GMPL report solve status and so we don't really need it here). I see that some of the statuses are set using the response code on `getResults` and others are reading it from `results.status` in this response. I think ideally we would have `SolverSolveEngine.status` just mirror the result of `result.status` and then use this (and not `solve_result_num`) in `SolverSolveEngine.getStatus()`. Is there a list of possible statuses that can be returned?

I'd also like to run it through some of our test problems to make sure everything is working (including the resuming support), but when I try to solve, I get an error that translating gmpl to lp is not possible:

```
Unexpected response code while creating a job: 400

The response was:
{
"status": 400,
"error": "Bad Request",
"message": "cannot translate from gmpl to lp",
"timestamp": "2016-11-21T22:06:55Z"
}
```

Is that to be expected?

### CODE REVIEW

**Comment by Tapanito** [comment]
@JackDunnNZ first of all, thank you very much for the amazing work you did!

> There's one more change I'd like to make before we land this, and that's to get rid of solve_result_num as an intermediate solve status (since this just how AMPL/GMPL report solve status and so we don't really need it here). 

You're absolutely right, I wasn't too sure how to approach this, thus I left it as it was. I'm more than happy to add the changes later on this week.

>  I see that some of the statuses are set using the response code on **`getResults`**

  Here we're making sure that the job (the submitted problem) hasn't failed due to some reason. The failure might be caused by failure to translate the problem between formats, or some internal error.
> and others are reading it from  **`results.status`**

  Here the status represents the status of the actual problem, which can be one of the:
  * Optimal
  * Infeasible 
  * Interrupted 

The **`SolverSolveEngine.getStatus()`** call returns in what solving state the job is, the value can be one of the following:
 * Queued     (the job has been created, but not started)
* Translating (the problem is being translated between formats)
* Started     (the problem is being solved)
* Starting    
* Completed  (the problem has been solved)
* Failed     (failed to solve the problem)
* Stopped    (the problem solving has been stopped by the user)

In short, if the job fails to complete, the **`SolverSolveEngine.getStatus()`** will return **`Failed`** and the **`SolverSolveEngine.getResults()`** will return response code **`400`**

However, we're still in the development stages, and any advice would be truly welcome!

> I'd also like to run it through some of our test problems to make sure everything is working (including the resuming support), but when I try to solve, I get an error that translating gmpl to lp is not possible

This has been fixed

Thank you very much for your work Jack!

**Comment by JackDunnNZ** [comment]
Thanks @Tapanito, that's very helpful. I've had a go at removing `solve_result_num`, can you check this logic makes sense?

I also realized the UserCache was not the right way to store the API key either because it would only persist the key for an hour IIRC. I switched it to use UserProperties instead which don't time out.

I'm just running through some more of our test problems now to see if I catch anything
